src/esp/sim/Simulator.cpp

```diff
@@ -1126,7 +1126,7 @@ agent::Agent::ptr Simulator::getAgent(const int agentId) {
 
 esp::sensor::Sensor::ptr Simulator::addSensorToObject(
     const int objectId,
-    esp::sensor::SensorSpec::ptr& sensorSpec) {
+    esp::sensor::SensorSpec::ptr sensorSpec) {
```

Comment:
Changed because there was trouble passing in derived classes.
Could not get type esp::sensor::SensorSpec::ptr& from esp::sensor::CameraSensorSpec::ptr

Reply:
However, this bugs clang-tidy because SensorSpec is not modified in addSensorToObject(). I did some googling but was not sure how to fix this

Reply:
I did not catch this in the previous PR.

`::ptr` is already a pointer so you do not have to pass the value by reference.

Second, which is more important:
The API design in the `Simulator::addSensorToObject` is really needed to be improved.
- you do not need `const` for an atomic int.
- the operation within this function has nothing to do with the ownership. so you do not need to pass a smart pointer. 

Suggested change:
`esp::sensor::Sensor::ptr Simulator::addSensorToObject(int objectId, const esp::sensor::SensorSpec& sensorSPec)`

Please remember to fix it in another PR.

Reply:
OK. I made a mistake here. 
Here, you are creating sensors from the sensor spec so the ownership of the spec will be transferred. So you need the ptr.

Reply:
You will have to use `esp::sensor::SensorSpec::ptr&`. Cannot change it to `esp::sensor::SensorSpec::ptr`

You will fail the compiling? or just fail the clang-tidy?

Reply:
Ok thanks! I was looking into that but could not make it work so was about to ask. SensorSpec::create() returns a SensorSpec::ptr, so we will rarely have SensorSpec.

Reply:
`esp::sensor::SensorSpec::ptr&` fails compiling but passes clang-tidy
`esp::sensor::SensorSpec::ptr` passes compiling but fails clang-tidy
Not quite sure what to do here - do you have any suggestions?

Reply:
You mentioned this error:
Error was: Could not get type esp::sensor::SensorSpec::ptr& from esp::sensor::CameraSensorSpec::ptr

May I have the location? which file? on which line?

Reply:
Let me push the change, it was when initializing a CameraSensorSpec in SimTest

Reply:
{
	"resource": "/Users/auduong/habitat-sim/src/tests/SimTest.cpp",
	"owner": "C/C++",
	"code": "434",
	"severity": 8,
	"message": "a reference of type \"esp::sensor::SensorSpec::ptr &\" (not const-qualified) cannot be initialized with a value of type \"esp::sensor::CameraSensorSpec::ptr\"",
	"source": "C/C++",
	"startLineNumber": 646,
	"startColumn": 58,
	"endLineNumber": 646,
	"endColumn": 74
}

Reply:
```
esp::sensor::SensorSuite sensorSuite;
  auto objectSensorSpec = esp::sensor::CameraSensorSpec::create();
  objectSensorSpec->uuid = std::to_string(objectID);
  objectSensorSpec->position = {0, 0, 0};
  objectSensorSpec->orientation = {0, 0, 0};
  objectSensorSpec->resolution = {128, 128};
  sensorSuite.add(simulator->addSensorToObject(objectID, objectSensorSpec));
```

Reply:
I need to go. can you try something like pointer_cast?

Reply:
Ok no problem, thanks for the help!